File: src/bmsx/rompack/source.ts
import type { asset_id, asset_type, CartridgeIndex, CartridgeLayerId, RomAsset } from './format';

export type RomSourceLayer = {
	id: CartridgeLayerId;
	index: CartridgeIndex;
	payload: Uint8Array;
};

export interface RawRomSource {
	getEntry(id: asset_id): RomAsset | null;
	getEntryByPath(path: string): RomAsset | null;
	getBytes(entry: RomAsset): Uint8Array;
	getBytesView(entry: RomAsset): Uint8Array;
	list(type?: asset_type): RomAsset[];
}

export class RomSourceStack implements RawRomSource {
	private readonly layers: RomSourceLayer[];
	private readonly idMaps: Map<string, RomAsset>[];
	private readonly pathMaps: Map<string, RomAsset>[];
	private readonly payloads: Partial<Record<CartridgeLayerId, Uint8Array>>;

	public constructor(layers: RomSourceLayer[]) {
		this.layers = layers;
		this.idMaps = layers.map(layer => {
			const map = new Map<string, RomAsset>();
			for (const entry of layer.index.entries) {
				map.set(entry.resid, entry);
			}
			return map;
		});
		this.pathMaps = layers.map(layer => {
			const map = new Map<string, RomAsset>();
			for (const entry of layer.index.entries) {
				if (!entry.source_path) {
					continue;
				}
				map.set(entry.source_path, entry);
			}
			return map;
		});
		const payloads: Partial<Record<CartridgeLayerId, Uint8Array>> = {};
		for (const layer of layers) {
			payloads[layer.id] = layer.payload;
		}
		this.payloads = payloads;
	}

	public getEntry(id: asset_id): RomAsset | null {
		for (let i = 0; i < this.layers.length; i++) {
			const asset = this.idMaps[i].get(id);
			if (!asset) {
				continue;
			}
			if (asset.op === 'delete') {
				return null;
			}
			return this.attachPayloadId(asset, this.layers[i].id);
		}
		return null;
	}

	public getEntryByPath(path: string): RomAsset | null {
		for (let i = 0; i < this.layers.length; i++) {
			const asset = this.pathMaps[i].get(path);
			if (!asset) {
				continue;
			}
			if (asset.op === 'delete') {
				return null;
			}
			return this.attachPayloadId(asset, this.layers[i].id);
		}
		return null;
	}

	public list(type?: asset_type): RomAsset[] {
		const resolved = new Map<string, RomAsset>();
		const blocked = new Set<string>();
		for (let layerIndex = 0; layerIndex < this.layers.length; layerIndex++) {
			const layer = this.layers[layerIndex];
			for (const entry of layer.index.entries) {
				if (type && entry.type !== type) {
					continue;
				}
				const id = entry.resid;
				if (blocked.has(id)) {
					continue;
				}
				if (entry.op === 'delete') {
					blocked.add(id);
					resolved.delete(id);
					continue;
				}
				resolved.set(id, this.attachPayloadId(entry, layer.id));
				blocked.add(id);
			}
		}
		return Array.from(resolved.values());
	}

	public getBytes(entry: RomAsset): Uint8Array {
		const payload = this.payloads[entry.payload_id];
		return payload.slice(entry.start, entry.end);
	}

	public getBytesView(entry: RomAsset): Uint8Array {
		const payload = this.payloads[entry.payload_id];
		return payload.subarray(entry.start, entry.end);
	}

	private attachPayloadId(asset: RomAsset, payloadId: CartridgeLayerId): RomAsset {
		if (asset.payload_id === payloadId) {
			return asset;
		}
		return { ...asset, payload_id: payloadId };
	}
}

File: src/bmsx_cpp/rompack/source.cpp
#include "rompack/source.h"

#include <unordered_set>


namespace bmsx {
namespace {

bool romEntryDeletes(const RomSourceEntry& entry) {
	return entry.rom.op.has_value() && *entry.rom.op == "delete";
}

const std::vector<u8>& payloadForEntry(const std::unordered_map<std::string, const std::vector<u8>*>& payloads, const RomAssetInfo& entry) {
	const auto it = payloads.find(*entry.payloadId);
	return *it->second;
}

} // namespace

RomSourceStack::RomSourceStack(std::vector<RomSourceLayer> layers)
	: m_layers(std::move(layers)) {
	m_idMaps.reserve(m_layers.size());
	m_pathMaps.reserve(m_layers.size());
	for (const RomSourceLayer& layer : m_layers) {
		std::unordered_map<std::string, size_t> idMap;
		std::unordered_map<std::string, size_t> pathMap;
		idMap.reserve(layer.index.entries.size());
		pathMap.reserve(layer.index.entries.size());
		for (size_t index = 0; index < layer.index.entries.size(); ++index) {
			const RomSourceEntry& entry = layer.index.entries[index];
			idMap[entry.resid] = index;
			if (entry.rom.sourcePath.has_value()) {
				pathMap[*entry.rom.sourcePath] = index;
			}
		}
		m_idMaps.push_back(std::move(idMap));
		m_pathMaps.push_back(std::move(pathMap));
		m_payloads[cartridgeLayerIdString(layer.id)] = layer.payload;
	}
}

std::optional<RomSourceEntry> RomSourceStack::getEntry(std::string_view id) const {
	for (size_t layerIndex = 0; layerIndex < m_layers.size(); ++layerIndex) {
		const auto assetIt = m_idMaps[layerIndex].find(std::string(id));
		if (assetIt == m_idMaps[layerIndex].end()) {
			continue;
		}
		const RomSourceEntry& asset = m_layers[layerIndex].index.entries[assetIt->second];
		if (romEntryDeletes(asset)) {
			return std::nullopt;
		}
		return attachPayloadId(asset, m_layers[layerIndex].id);
	}
	return std::nullopt;
}

std::optional<RomSourceEntry> RomSourceStack::getEntryByPath(std::string_view path) const {
	for (size_t layerIndex = 0; layerIndex < m_layers.size(); ++layerIndex) {
		const auto assetIt = m_pathMaps[layerIndex].find(std::string(path));
		if (assetIt == m_pathMaps[layerIndex].end()) {
			continue;
		}
		const RomSourceEntry& asset = m_layers[layerIndex].index.entries[assetIt->second];
		if (romEntryDeletes(asset)) {
			return std::nullopt;
		}
		return attachPayloadId(asset, m_layers[layerIndex].id);
	}
	return std::nullopt;
}

std::vector<RomSourceEntry> RomSourceStack::list(std::optional<std::string_view> type) const {
	std::vector<RomSourceEntry> out;
	std::unordered_set<std::string> blocked;
	for (const RomSourceLayer& layer : m_layers) {
		for (const RomSourceEntry& entry : layer.index.entries) {
			if (type.has_value() && entry.rom.type != *type) {
				continue;
			}
			if (blocked.find(entry.resid) != blocked.end()) {
				continue;
			}
			if (romEntryDeletes(entry)) {
				blocked.insert(entry.resid);
				continue;
			}
			out.push_back(attachPayloadId(entry, layer.id));
			blocked.insert(entry.resid);
		}
	}
	return out;
}

std::vector<u8> RomSourceStack::getBytes(const RomSourceEntry& entry) const {
	return getBytes(entry.rom);
}

std::vector<u8> RomSourceStack::getBytes(const RomAssetInfo& entry) const {
	std::span<const u8> view = getBytesView(entry);
	return std::vector<u8>(view.begin(), view.end());
}

std::span<const u8> RomSourceStack::getBytesView(const RomSourceEntry& entry) const {
	return getBytesView(entry.rom);
}

std::span<const u8> RomSourceStack::getBytesView(const RomAssetInfo& entry) const {
	const std::vector<u8>& payload = payloadForEntry(m_payloads, entry);
	return std::span<const u8>(payload.data() + static_cast<size_t>(*entry.start), static_cast<size_t>(*entry.end - *entry.start));
}

RomSourceEntry RomSourceStack::attachPayloadId(const RomSourceEntry& asset, CartridgeLayerId payloadId) const {
	RomSourceEntry copy = asset;
	copy.rom.payloadId = cartridgeLayerIdString(payloadId);
	return copy;
}

const char* cartridgeLayerIdName(CartridgeLayerId id) {
	switch (id) {
		case CartridgeLayerId::System: return "system";
		case CartridgeLayerId::Cart: return "cart";
		case CartridgeLayerId::Overlay: return "overlay";
	}
	throw BMSX_RUNTIME_ERROR("Invalid cartridge layer id.");
}

std::string cartridgeLayerIdString(CartridgeLayerId id) {
	return cartridgeLayerIdName(id);
}

CartridgeLayerId cartridgeLayerIdFromString(std::string_view id) {
	if (id == "system") return CartridgeLayerId::System;
	if (id == "overlay") return CartridgeLayerId::Overlay;
	return CartridgeLayerId::Cart;
}

const u8* romSourceLayerBytes(const RomSourceLayer& layer, const RomAssetInfo& entry) {
	return layer.payload->data() + static_cast<size_t>(*entry.start);
}

size_t romSourceLayerByteLength(const RomAssetInfo& entry) {
	return static_cast<size_t>(*entry.end - *entry.start);
}

} // namespace bmsx
